File: momsfab/doc_events/material_request.py
import frappe, json
from frappe.model.mapper import get_mapped_doc

@frappe.whitelist()
def on_submit_mr(doc, method):
    for i in doc.budget_bom_reference:
        if i.budget_bom:

            frappe.db.sql(""" UPDATE `tabBudget BOM` SET status='To Purchase Order' WHERE name=%s """, i.budget_bom)
            frappe.db.commit()
@frappe.whitelist()
def get_budget_bom(doctype,target,e,r,t,filter):
    print("FILTEEEEEEEEEEEEEEEER")
    print(filter)
    print(tuple(filter['data']))
    condition = ""
    if len(filter['data']) == 1:
        condition += " and name!='{0}'".format(filter['data'][0])

    if len(filter['data']) > 1:
        condition += " and name not in {0}".format(tuple(filter['data']))

    query = """ SELECT opportunity as name, name as budget_bom, customer_name FROM `tabBudget BOM` WHERE status='To Material Request' and docstatus=1 {0}""".format(condition)
    opportunities = frappe.db.sql(query,as_dict=1)
    return opportunities


def validate_mr(doc, method):

    for i in doc.budget_bom_reference:
        if i.budget_bom:
            for ii in doc.items:
                item = frappe.get_doc("Item", ii.item_code)
                if not ii.uom:
                    ii.uom = item.stock_uom
                    ii.description = item.item_name

                if ii.budget_bom_raw_material:
                    rate = frappe.db.sql(""" SELECT * from `tabBudget BOM Details` WHERE name=%s""", ii.budget_bom_raw_material, as_dict=1)
                    rate1 = frappe.db.sql(""" SELECT * from `tabBudget BOM Details Engineering` WHERE name=%s""", ii.budget_bom_raw_material, as_dict=1)
                    rate2 = frappe.db.sql(""" SELECT * from `tabBudget BOM Details Pipe Estimation` WHERE name=%s""", ii.budget_bom_raw_material, as_dict=1)

                    if len(rate) > 0:
                        ii.budget_bom_rate =  ii.rate if rate[0].rate == 0 else rate[0].rate

                    elif len(rate1) > 0:
                        ii.budget_bom_rate =  ii.rate if rate1[0].rate == 0 else rate1[0].rate

                    elif len(rate2) > 0:
                        ii.budget_bom_rate =  ii.rate if rate2[0].rate == 0 else rate2[0].rate

def cancel_mr(doc, method):
    for i in doc.budget_bom_reference:
        if i.budget_bom:
            frappe.db.sql(""" UPDATE `tabBudget BOM` SET status=%s WHERE name=%s""", ('To Material Request', i.budget_bom))
            frappe.db.commit()


@frappe.whitelist()
def get_bb(mr):
    data = json.loads(mr)
    items = []
    bbs = []
    for x in data:
        bb_document = frappe.get_doc("Budget BOM", x)
        if bb_document.type == "Customer":
            frappe.throw("MR cannot be generated for Budget BOM " + str(x))

        bb_doc = get_mapped_doc("Budget BOM", x, {
            "Budget BOM": {
                "doctype": "Material Request",
                "validation": {
                    "docstatus": ["=", 1]
                }
            },
            "Budget BOM Details": {
                "doctype": "Material Request Item",
                "field_map": {
                    "name": "budget_bom_raw_material",
                    "weight_of_sheet": "qty"

                }
            },
            "Budget BOM Details Engineering": {
                "doctype": "Material Request Item",
                "field_map": {
                    "name": "budget_bom_raw_material",
                    "weight_of_sheet": "qty"

                }
            },
            "Budget BOM Details Pipe Estimation": {
                "doctype": "Material Request Item",
                "field_map": {
                    "name": "budget_bom_raw_material",
                    "production_qty": "qty"
                }
            }
        })
        so = frappe.db.sql(""" SELECT SOI.*, SO.transaction_date FROM `tabSales Order` SO 
                              INNER JOIN `tabSales Order Item` SOI ON SOI.parent = SO.name 
                              INNER JOIN `tabBudget BOM References` BBR ON BBR.parent = SO.name  
                              WHERE BBR.budget_bom = %s
                              """,x,as_dict=1)

        qty = check_qty(bb_document.fg_sellable_bom_details[0].item_code, so)

        for xx in bb_doc.items:
            xx.qty = xx.qty * qty
            xx.schedule_date = bb_document.posting_date

            items.append(xx)

        items = consolidate_items(items)
        bbs.append({
            "budget_bom": x
        })

    return items, bbs

def check_qty(item_code, items):
    for i in items:
        if i.item_code == item_code:
            return i.qty
    return 1
def consolidate_items(items):
    c_items = []
    for i in items:
        add = False
        for x in c_items:
            if i.item_code == x.item_code:
                if 'type' in i.__dict__ and i.type:
                    if i.type == 'Deletion':
                        x.qty -= i.qty
                    else:
                        x.qty += i.qty
                else:
                    x.qty += i.qty
                add = True
        if not add:
            c_items.append(i)

    return c_items

File: momsfab/public/js/material_request.js
frappe.ui.form.on("Budget BOM References", {
    budget_bom: function (frm, cdt, cdn) {
        var d = locals[cdt][cdn]

        if(d.budget_bom){
            frappe.db.get_doc("Budget BOM",d.budget_bom)
                .then(doc => {
                var field = doc.type === 'Sheet Estimation' ? "sheet_estimation" :
                                doc.type === 'Engineering Estimation' ? "engineering_estimation" :
                                    doc.type === 'Pipe Estimation' ? "pipe_estimation": ""

                for(var i=0;i<doc[field].length;i+=1) {
                    if (!check_items(doc[field][i], cur_frm)) {
                        cur_frm.add_child("items", {
                            item_code: doc[field][i].item_code,
                            item_name: doc[field][i].item_name,
                            description: doc[field][i].item_name,
                            qty: doc[field][i].qty,
                            uom: doc[field][i].stock_uom,
                            conversion_factor:1,
                            stock_uom: doc[field][i].stock_uom,
                            rate: doc[field][i].rate,
                            budget_bom_rate: doc[field][i].rate,
                        })
                        cur_frm.refresh_field("items")
                    }
                }

            })
        }

    }
})
frappe.ui.form.on("Material Request", {
    refresh: function (frm, cdt, cdn) {
    if(cur_frm.is_new()){
         cur_frm.add_custom_button(__('Budget BOM'),
				function() {
            var budget_boms = []
            if(cur_frm.doc.budget_bom_reference){
                 budget_boms = Array.from(cur_frm.doc.budget_bom_reference, x => "budget_bom" in x ? x.budget_bom:"")
            }
                    var query_args = {
                        query:"momsfab.doc_events.material_request.get_budget_bom",
                        filters: {data: budget_boms}
                    }
					 var d = new frappe.ui.form.MultiSelectDialog({
                            doctype: "Opportunity",
                            target: cur_frm,
                            setters: [
                                {
                                    label: "Budget BOM",
                                    fieldname: "budget_bom",
                                    fieldtype: "Link",
                                    options: "Budget BOM",
                                    default:  undefined
                                },
                                 {
                                    label: "Customer",
                                    fieldname: "customer_name",
                                    fieldtype: "Data",
                                    default: cur_frm.doc.party_name || undefined
                                }
                            ],
                            add_filters_group: 0,
                            date_field: "posting_date",
                            get_query() {
                                return query_args
                            },
                            action(selections) {
                                fetch_boms(cur_frm, selections)
                                d.dialog.hide()
                            }
                        });
				}, __("Get Items From"), "btn-default");
    }


    }
})

function fetch_boms(cur_frm, selections) {
    if(cur_frm.doc.items.length> 0 && !cur_frm.doc.items[0].item_code){
        cur_frm.clear_table("items")
        cur_frm.refresh_field("items")
    }
   if(selections.length > 0){
        frappe.call({
            method: "momsfab.doc_events.material_request.get_bb",
            args: {
                mr: selections
            },
            callback: function (r) {
                    for(var x=0;x<r.message[0].length;x+=1){
                        cur_frm.add_child("items",r.message[0][x])
                        cur_frm.refresh_fields("items")
                    }
                    for(var x=0;x<r.message[1].length;x+=1){
                        cur_frm.add_child("budget_bom_reference",r.message[1][x])
                        cur_frm.refresh_fields("budget_bom_reference")
                    }
            }
        })
    }
}
function check_items(item, cur_frm) {
        for(var x=0;x<cur_frm.doc.items.length;x+=1){
            var item_row = cur_frm.doc.items[x]
            if(item_row.item_code === item.item_code){
                item_row.qty += item.qty
                cur_frm.refresh_field("items")
                return true
            }
        }
        return false
}